datalake/src/main/java/com/sequenceiq/datalake/controller/DatalakeEventController.java

```diff
@@ -0,0 +1,124 @@
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.StreamingOutput;
+
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+
+import com.sequenceiq.authorization.utils.EventAuthorizationDto;
+import com.sequenceiq.authorization.utils.EventAuthorizationUtils;
+import com.sequenceiq.cloudbreak.auth.crn.Crn;
+import com.sequenceiq.cloudbreak.auth.crn.CrnParseException;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.common.json.JsonUtil;
+import com.sequenceiq.cloudbreak.structuredevent.event.StructuredEventType;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredEvent;
+import com.sequenceiq.cloudbreak.structuredevent.service.db.CDPStructuredEventDBService;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.repository.SdxClusterRepository;
+import com.sequenceiq.sdx.api.endpoint.DatalakeEventEndpoint;
+
+public class DatalakeEventController implements DatalakeEventEndpoint {
+
+    @Inject
+    private CDPStructuredEventDBService cdpStructuredEventDBService;
+
+    @Inject
+    private EventAuthorizationUtils eventAuthorizationUtils;
+
+    @Inject
+    private SdxClusterRepository sdxClusterRepository;
+
+    /**
+     * Retrieves audit events for all datalakes that have been associated with the provided environment CRN.
+     *
+     * @param resourceCrn an environment CRN
+     * @param types types of structured events to retrieve
+     * @param page the page index
+     * @param size the size of the page to retrieve
+     * @return structured events for all datalakes related to the resource CRN of an environment
+     */
+    @Override
+    public List<CDPStructuredEvent> getAuditEvents(String resourceCrn, List<StructuredEventType> types, Integer page, Integer size) {
+        PageRequest pageable = PageRequest.of(page, size, Sort.by("timestamp").descending());
+        // datalake crns related to the provided environment CRN.
+        List<String> datalakeCrns = getRelatedDatalakeCrns(resourceCrn);
+
+        // actually retrieve events from DB
+        List<CDPStructuredEvent> events = cdpStructuredEventDBService.getPagedEventsOfResources(types, datalakeCrns, pageable).getContent();
+        if (events.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        // todo: make sure this permission check actually ensures unauthorized users can't access the info
+        eventAuthorizationUtils.checkPermissionBasedOnResourceTypeAndCrn(collectDtosFromEvents(events));
+
+        return events;
+    }
```

Comment:
How about the below implementation. This implementation gets the events from cdpStructuredEventV1Endpoint and merges the result. Doesn't this work?

```
    @Override
    public List<CDPStructuredEvent> getAuditEvents(String resourceCrn, List<StructuredEventType> types, Integer page, Integer size) {
        PageRequest pageable = PageRequest.of(page, size, Sort.by("timestamp").descending());
        // datalake crns related to the provided environment CRN.
        List<String> datalakeCrns = getRelatedDatalakeCrns(resourceCrn);

        // actually retrieve events from DB
        List<CDPStructuredEvent> dlEvents = cdpStructuredEventDBService.getPagedEventsOfResources(types, datalakeCrns, pageable).getContent();
        if (dlEvents.isEmpty()) {
            return Collections.emptyList();
        }

        List<CDPStructuredEvent> cbEvents =  cdpStructuredEventV1Endpoint.getAuditEvents(datalakeCrns.get(0), types, page, size);

        dlEvents.addAll(cbEvents);
        dlEvents = dlEvents.stream().sorted(Comparator.comparingLong(f -> f.getOperation().getTimestamp())).collect(Collectors.toList());

        // todo: make sure this permission check actually ensures unauthorized users can't access the info
        eventAuthorizationUtils.checkPermissionBasedOnResourceTypeAndCrn(collectDtosFromEvents(dlEvents));

        return dlEvents;
    }
```

Reply:
Kalyan and I discussed this, the only issue with this approach is that the page values are allowed to differ pretty wildly.

Using the default page values of `(0, 100)` means that we'll be getting the "beginning" of each of the events, so merging them together should at least make sense.

There's a possibility of pretty big gaps in the returned stream though, as the pages differ more and more from the timestamp values.

Here's a possible diff:
```diff
diff --git a/datalake/src/main/java/com/sequenceiq/datalake/controller/DatalakeEventController.java b/datalake/src/main/java/com/sequenceiq/datalake/controller/DatalakeEventController.java
index 5fa3059f61..d12fbe1a4b 100644
--- a/datalake/src/main/java/com/sequenceiq/datalake/controller/DatalakeEventController.java
+++ b/datalake/src/main/java/com/sequenceiq/datalake/controller/DatalakeEventController.java
@@ -5,8 +5,10 @@ import static java.util.stream.Collectors.toSet;
 
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 
@@ -29,6 +31,7 @@ import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredEvent;
 import com.sequenceiq.cloudbreak.structuredevent.service.db.CDPStructuredEventDBService;
 import com.sequenceiq.datalake.entity.SdxCluster;
 import com.sequenceiq.datalake.repository.SdxClusterRepository;
+import com.sequenceiq.environment.client.EnvironmentClient;
 import com.sequenceiq.sdx.api.endpoint.DatalakeEventEndpoint;
 
 public class DatalakeEventController implements DatalakeEventEndpoint {
@@ -42,6 +45,9 @@ public class DatalakeEventController implements DatalakeEventEndpoint {
     @Inject
     private SdxClusterRepository sdxClusterRepository;
 
+    @Inject
+    private EnvironmentClient environmentclient;
+
     /**
      * Retrieves audit events for all datalakes that have been associated with the provided environment CRN.
      *
@@ -58,15 +64,22 @@ public class DatalakeEventController implements DatalakeEventEndpoint {
         List<String> datalakeCrns = getRelatedDatalakeCrns(resourceCrn);
 
         // actually retrieve events from DB
-        List<CDPStructuredEvent> events = cdpStructuredEventDBService.getPagedEventsOfResources(types, datalakeCrns, pageable).getContent();
-        if (events.isEmpty()) {
+        List<CDPStructuredEvent> dlEvents = cdpStructuredEventDBService.getPagedEventsOfResources(types, datalakeCrns, pageable).getContent();
+        if (dlEvents.isEmpty()) {
             return Collections.emptyList();
         }
 
+        List<CDPStructuredEvent> cbEvents = environmentclient.structuredEventsV1Endpoint().getAuditEvents(datalakeCrns.get(0), types, page, size);
+
+        dlEvents.addAll(cbEvents);
+        dlEvents = dlEvents.stream()
+                .sorted(Comparator.comparingLong(f -> f.getOperation().getTimestamp()))
+                .collect(Collectors.toList());
+
         // todo: make sure this permission check actually ensures unauthorized users can't access the info
-        eventAuthorizationUtils.checkPermissionBasedOnResourceTypeAndCrn(collectDtosFromEvents(events));
+        eventAuthorizationUtils.checkPermissionBasedOnResourceTypeAndCrn(collectDtosFromEvents(dlEvents));
 
-        return events;
+        return dlEvents;
     }
 
     private String getAccountId(String crnString) {
```